src/module_3/assign_3.md

```diff
@@ -0,0 +1,980 @@
+      <td>31.017468</td>
+      <td>30.0</td>
+      <td>39.0</td>
+      <td>35.641908</td>
+    </tr>
+    <tr>
+      <th>2860261</th>
+      <td>1</td>
+      <td>34304342851716</td>
+      <td>2829221494916</td>
+      <td>3784403714180</td>
+      <td>10120</td>
+      <td>1.604102e+09</td>
+      <td>2</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>...</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>2.0</td>
+      <td>33.0</td>
+      <td>47.0</td>
+      <td>31.017468</td>
+      <td>30.0</td>
+      <td>39.0</td>
+      <td>35.641908</td>
+    </tr>
+    <tr>
+      <th>2863549</th>
+      <td>1</td>
+      <td>33826433859716</td>
+      <td>2823601127556</td>
+      <td>3469897564292</td>
+      <td>20585</td>
+      <td>1.603411e+09</td>
+      <td>2</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>...</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>2.0</td>
+      <td>33.0</td>
+      <td>82.0</td>
+      <td>44.243644</td>
+      <td>30.0</td>
+      <td>38.5</td>
+      <td>31.832738</td>
+    </tr>
+    <tr>
+      <th>2863654</th>
+      <td>1</td>
+      <td>33826433859716</td>
+      <td>2829871546500</td>
+      <td>3771764834436</td>
+      <td>20585</td>
+      <td>1.604189e+09</td>
+      <td>3</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>...</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>2.0</td>
+      <td>33.0</td>
+      <td>82.0</td>
+      <td>44.243644</td>
+      <td>30.0</td>
+      <td>38.5</td>
+      <td>31.832738</td>
+    </tr>
+  </tbody>
+</table>
+<p>42868 rows × 26 columns</p>
+</div>
+
+
+
+
+```python
+balanced_train_data_x = balanced_df.iloc[:,4:]
+balanced_train_data_y = balanced_df.iloc[:,0]
+
+#Logistic regression with balanced training data
+
+param_grid = {
+    'C': [10 ** i for i in range(-4,2)],
+    'penalty': ['l2']
+}
+
+log_model_balanced = LogisticRegression(solver='lbfgs',n_jobs=-1, class_weight='balanced')
+
+grid_search_balanced = GridSearchCV(log_model_balanced, param_grid, cv=5, scoring='accuracy').fit(balanced_train_data_x,balanced_train_data_y)
+
+print(grid_search_balanced.best_params_)
+
+y_train_pred_balanced = grid_search_balanced.predict(balanced_train_data_x)
+y_train_prob_balanced = grid_search_balanced.predict_proba(balanced_train_data_x)[:,1]
+
+draw_roc_prc(balanced_train_data_y, y_train_prob_balanced,"With regularization and a balanced dataset")
+print(metrics.confusion_matrix(balanced_train_data_y,y_train_pred_balanced))
+
+
+```
+
+    {'C': 0.0001, 'penalty': 'l2'}
+
+
+
+    
+![png](assign_3_files/assign_3_18_1.png)
+    
+
+
+    [[21434     0]
+     [21434     0]]
+
+
+
+```python
+threshold_balanced = best_threshold(balanced_train_data_y, y_train_prob_balanced)
+```
+
+    Best Threshold: 0.4999481480929332
+
+
+
+```python
+#Linear regression model with Reguarization and F1 scoring
```

Comment:
The process of training is repeated over and over, could and should be extracted on a function. That function will be reused later to write production-ready code.

Reply:
Implemented a function for this as well!